### PR TITLE
Implement EventListenerOptions for EventTarget

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
It still needs to update tests and confirm that the code is the one expected. Reopening of #12952.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/15803)

### CODE REVIEW

**Comment by nox** [comment]
It's not. In https://github.com/servo/servo/pull/12952#issuecomment-241336248 @Ms2ger mentions we only want to support the `capture` key for now, given we don't implement the features related to the other keys.

**Comment by GuillaumeGomez** [comment]
So you want me to comment all `::Boolean` cases for now?

**Comment by nox** [comment]
I don't understand the question.

I want:

```webidl
dictionary EventListenerOptions {
  boolean capture = false;
};

dictionary AddEventListenerOptions : EventListenerOptions {
  // boolean passive = false;
  // boolean once = false;
};
```

**Comment by GuillaumeGomez** [comment]
Ok, makes more sense. I thought you were only speaking of the usage, not the declaration.

**Comment by Ms2ger** [comment]
@GuillaumeGomez ping

**Comment by GuillaumeGomez** [comment]
Sorry, was busy because of rustdoc. Will finish it this week-end.

**Comment by GuillaumeGomez** [comment]
Updated.

File: components/script/dom/eventtarget.rs [hunk 1]
```diff
@@ -225,7 +228,11 @@ impl CompiledEventListener {
 /// A listener in a collection of event listeners.
 struct EventListenerEntry {
     phase: ListenerPhase,
-    listener: EventListenerType
+    listener: EventListenerType,
+    capture: bool,
+    passive: bool,
+    once: bool,
+    removed: bool,
```
Comment (nox) [inline components/script/dom/eventtarget.rs]: Don't add fields for which we don't support the corresponding thing, so just add `capture`.

**Comment by GuillaumeGomez** [comment]
Updated.

File: components/script/dom/eventtarget.rs [hunk 2]
```diff
@@ -225,7 +228,11 @@ impl CompiledEventListener {
 /// A listener in a collection of event listeners.
 struct EventListenerEntry {
     phase: ListenerPhase,
-    listener: EventListenerType
+    listener: EventListenerType,
+    capture: bool,
+    // passive: bool,
+    // once: bool,
+    // removed: bool,
```
Comment (nox) [inline components/script/dom/eventtarget.rs]: Remove that.

File: components/script/dom/eventtarget.rs [hunk 3]
```diff
@@ -337,6 +344,10 @@ impl EventTarget {
                     entries.push(EventListenerEntry {
                         phase: ListenerPhase::Bubbling,
                         listener: EventListenerType::Inline(listener),
+                        capture: false,
+                        // passive: false,
+                        // once: false,
+                        // removed: false,
```
Comment (nox) [inline components/script/dom/eventtarget.rs]: Remove that.

File: components/script/dom/eventtarget.rs [hunk 4]
```diff
@@ -550,46 +561,63 @@ impl EventTargetMethods for EventTarget {
...
+                        options: AddEventListenerOptionsOrBoolean) {
+        if let Some(listener) = listener {
+            let (capture, _, _) = match options {
+                AddEventListenerOptionsOrBoolean::AddEventListenerOptions(
+                    AddEventListenerOptions { .. }) => {
+                    // (false, once, passive)
+                    (false, false, false)
+                }
+                AddEventListenerOptionsOrBoolean::Boolean(o) => (o, false, false),
+            };
```
Comment (nox) [inline components/script/dom/eventtarget.rs]:
```rust
let capture = match options {
    AddEventListenerOptionsOrBoolean::AddEventListenerOptions(options) => options.capture,
    AddEventListenerOptionsOrBoolean::Boolean(capture) => capture,
};

File: components/script/dom/eventtarget.rs [hunk 5]
```diff
@@ -550,46 +561,63 @@ impl EventTargetMethods for EventTarget {
...
-            Some(l) => l,
-            None => return,
-        };
-        let mut handlers = self.handlers.borrow_mut();
-        let entry = match handlers.entry(Atom::from(ty)) {
-            Occupied(entry) => entry.into_mut(),
-            Vacant(entry) => entry.insert(EventListeners(vec!())),
-        };
+                        options: AddEventListenerOptionsOrBoolean) {
+        if let Some(listener) = listener {
```
Comment (nox) [inline components/script/dom/eventtarget.rs]: Why that change? Please early return just like before.

File: components/script/dom/eventtarget.rs [hunk 6]
```diff
@@ -550,46 +561,63 @@ impl EventTargetMethods for EventTarget {
...
+            };
+
+            let phase = if capture { ListenerPhase::Capturing } else { ListenerPhase::Bubbling };
+            let new_entry = EventListenerEntry {
+                phase: phase,
+                listener: EventListenerType::Additive(listener),
+                capture: capture,
+                // passive: passive,
+                // once: once,
+                // removed: false,
```
Comment (nox) [inline components/script/dom/eventtarget.rs]: Remove.

File: components/script/dom/eventtarget.rs [hunk 7]
```diff
@@ -550,46 +561,63 @@ impl EventTargetMethods for EventTarget {
...
-        };
-        let mut handlers = self.handlers.borrow_mut();
-        let entry = handlers.get_mut(&Atom::from(ty));
-        for entry in entry {
-            let phase = if capture { ListenerPhase::Capturing } else { ListenerPhase::Bubbling };
-            let old_entry = EventListenerEntry {
-                phase: phase,
-                listener: EventListenerType::Additive(listener.clone())
+                           options: EventListenerOptionsOrBoolean) {
+        if let Some(ref listener) = listener {
```
Comment (nox) [inline components/script/dom/eventtarget.rs]: Early return just like before.

File: components/script/dom/eventtarget.rs [hunk 8]
```diff
@@ -225,7 +228,11 @@ impl CompiledEventListener {
 /// A listener in a collection of event listeners.
 struct EventListenerEntry {
     phase: ListenerPhase,
-    listener: EventListenerType
+    listener: EventListenerType,
+    capture: bool,
```
Comment (nox) [inline components/script/dom/eventtarget.rs]: This new field is useless for now, given `phase` already covers it, please remove.

File: components/script/dom/eventtarget.rs [hunk 9]
```diff
@@ -550,46 +561,63 @@ impl EventTargetMethods for EventTarget {
...
+            let entry = handlers.get_mut(&Atom::from(ty));
+            for entry in entry {
+                let phase = if capture { ListenerPhase::Capturing } else { ListenerPhase::Bubbling };
+                let old_entry = EventListenerEntry {
+                    phase: phase,
+                    listener: EventListenerType::Additive(listener.clone()),
+                    capture: capture,
+                    // passive: false,
+                    // once: false,
+                    // removed: false,
```
Comment (nox) [inline components/script/dom/eventtarget.rs]: Remove that.

**Review by nox — CHANGES_REQUESTED** [review]

**Comment by GuillaumeGomez** [comment]
Updated.

**Review by nox — APPROVED** [review]

**Comment by jdm** [comment]
All tests are failing with errors like:
```
  │ ERROR:script::dom::bindings::error: Error at http://web-platform.test:8000/resources/testharness.js:663:9 Value not an object.
  │ ERROR:script::dom::bindings::error: Error at http://web-platform.test:8000/resources/testharnessreport.js:8:1 setup is not defined
  └ ERROR:script::dom::bindings::error: Error at http://web-platform.test:8000/css-flexbox-1_dev/html/flexbox_computedstyle_flex-direction-column-reverse.htm:16:1 test is not defined
```

**Comment by jdm** [comment]
@GuillaumeGomez Are you planning to return to this?

**Comment by GuillaumeGomez** [comment]
Yes, I'm currently working on it (slowly but still).

**Comment by GuillaumeGomez** [comment]
Actually I'm completely stuck in here and I have no idea what change broke everything. If anyone could help, it'd very appreciated!

**Comment by nox** [comment]
Superseded by #18612. Thanks for your contribution!

**Pull request closed without merging**